Matching TPC-only tracks to TRD tracklets

martenole opened the pull request:
This is WIP and not yet ready to be merged!

Implemented changes:

- created TRD track interface for `o2::track::TrackParCov` type instead of `o2::dataformats::TrackTPCITS`. TRD tracks can be created from both TPC-only tracks and from ITS-TPC matched tracks
- in case of TPC-only tracks which are not CE-crossing and for which thus the vertex is not known a-priori the TRD trigger records are searched for compatible collision candidates. The track following is done for each collision and the track with the lowest chi2 is stored
- in order to not have to refit the tracks for each collision not the track is shifted in z but instead the tracklet is shifted before the update and the chi2 calculation

What still needs to be done:

- the winner track needs to be refitted if it is matched to tracklets from a different collision than its initial estimate
- the logic in the `TRDGlobalTrackingSpec` needs to be adapted such that both ITS-TPC tracks and TPC-only tracks or just one or the other can be used as input
- the TRD tracks need to get global track IDs which point to the seed (TPC or ITS and TPC tracks)
- disable debug output again

@davidrohr If I want to look at the output tracks of the tracking which is stored in a tree called `tracksTRD` in the file `trdtracks.root` and I do `tracksTRD->Draw("tracks.mNTracklets")` I get a complaint from root: 
`Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::track::TrackParametrizationWithError with content: [...]`
I have adjusted the `GPUTrackingLinkDef_O2.h` for the new interface. Before when I used `TrackTPCITS` (which also inherits from `TrackParCov`) the root interpreter did not complain. Do you have an idea what might be the reason?

martenole committed:
TRD tracking based on TPC-only tracks implemented

- created track interface for TrackParCov
- references to ITS and TPC track seeds are added as GlobalTrackID
- TRD tracking can be configured via CLI to use ITS-TPC tracks, TPC-only tracks or both as input
- ITS-TPC track time uncertainty accounted for by increasing sigZ2
- TRD matches are written into separate files to allow for running two separate workflows

martenole commented:
In principle this is ready (at least for review).

- the TODOs mentioned above are implemented except for the refit part, although this can also be added later
- the flag `--tracking-sources arg` defines the input sources. Valid args are `ITS-TPC`, `TPC` or `ITS-TPC,TPC`. The default is `all`, so both ITS-TPC and TPC-only tracks
- I still have the ROOT interpreter issue which I mentioned above. I don't know how to fix it

shahor02 commented on Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx:
```diff
@@ -79,21 +84,32 @@ void TRDGlobalTracking::init(InitContext& ic)
 
   mTracker->PrintSettings();
 
+  auto& elParam = o2::tpc::ParameterElectronics::Instance();
+  auto& gasParam = o2::tpc::ParameterGas::Instance();
+  mTPCTBinMUS = elParam.ZbinWidth;
+  mTPCVdrift = gasParam.DriftV;
+  mTracker->SetTPCVdrift(mTPCVdrift); // possibly a value from the CCDB should be used instead?
+
```
I think it is better to move this in e.g. `updateTimeDependentParams` method called from `run()`, since in the real run these parameters should be queried for particular TF timestamp. 
I've already started to do this in different workflows.

martenole replied:
Sure that I have changed. Although the proper CCDB interface is still needed. I guess the TPC time bin granularity will be constant, but vDrift changes regularly

shahor02 commented on GPU/GPUTracking/TRDTracking/GPUTRDInterfaces.h:
```diff
@@ -139,35 +148,29 @@ namespace gpu
-  trackInterface<o2::dataformats::TrackTPCITS>(const GPUTPCGMMergedTrack& trk)
+  trackInterface<o2::track::TrackParCov>() = default;
+  trackInterface<o2::track::TrackParCov>(const o2::track::TrackParCov& param) = delete;
+  trackInterface<o2::track::TrackParCov>(const o2::dataformats::TrackTPCITS& trkItsTpc) : o2::track::TrackParCov(trkItsTpc.getParamOut()), mTime(trkItsTpc.getTimeMUS().getTimeStamp()), mTimePrecise(true)
```
The ITS-TPC track can have up to 1 \mus uncertainty, don't you account for it? You could probably use symmetric mTimeAdd/SubMax equal to `trkItsTpc.getTimeMUS().getTimeStamp()` multiplied by some NSigma settable parameter. Then the flag `preciseTimeKnown` would lose its meaning and you would need to treat all tracks on the same footing, looking just on the time uncertainty.

martenole replied:
I do not yet account for it, no. For the time uncertainty can I also query the `TrackTPCITS::timeEst` which is of type `TimeStampWithError`? Then I could set mTime[Add/Sub]Max from that error value and don't need another configurable NSigma parameter.
How large is the (maximum) uncertainty for CE-crossing TPC tracks? At the moment I would not shift these in z, even if the track time differs from the trigger time.

davidrohr replied:
@martenole : For the TPC only tracks: The TrackTPC has members `mDeltaTFwd` and `mDeltaTBwd` which indicates the maximum possible time to be shifted. This is also the case for CE-crossing tracks.

martenole replied:
Ok, I will check then how large this gets. Because at the moment I shift the tracks only in case they are A/C-side only (and from the side I determine the shift direction). If they are CE crossing I guess I will simply check the dip angle to determine the shift direction.

shahor02 replied:
Yes, timeEst gives both the time and error(sigma).
The error depends on pt, from a few 10ns to a few 100. You don't need to shift the track to the boundaries, it is enough to shift to the central value, but you can account it as an extra error in z.

martenole replied:
Both for ITS-TPC tracks and for CE-crossing TPC-only tracks I don't shift the tracks to the boundaries, but I increase the track uncertainty in Z instead.
Probably I need to rebase this PR again after https://github.com/AliceO2Group/AliceO2/pull/5994 is merged due to a conflict in `Detectors/TRD/workflow/CMakeLists.txt`

shahor02 reviewed (changes requested):
Thanks @martenole! please see 2 comments below

davidrohr commented:
@martenole : Could you resolve the conflicts and fix the compilation and update the PR :)
Sorry that I broke it :)

davidrohr commented:
It currently fails since it doesn't know TrackTPC and TrackTPCITS on the GPU. TrackTPCITS is actually a problem, since I believe the GlobalTrackId container will not work out of the box. Will try to have a look at that later. We need that anyway.
Perhaps I'd just hide the constructor (and related code) for TrackTPCITS with a GPUCA_GPUCODE #ifndef.

I am a bit puzzled why the TrackTPC is incomplete on the GPU. At least, TrackTPC should compile on the GPU correctly already.  Perhaps, have you hidden the header on the GPU?

> Probably I need to rebase this PR again after #5994 is merged due to a conflict in `Detectors/TRD/workflow/CMakeLists.txt`
Hope to merge that soon, then you can rebase.

shahor02 commented:
@davidrohr In the AbstractRef, which is the base class of GlobalTrackID, I specially avoided using bitsets, hoping that this is enough for it to work also on GPU. Are there other problems?

davidrohr commented:
> @davidrohr In the AbstractRef, which is the base class of GlobalTrackID, I specially avoided using bitsets, hoping that this is enough for it to work also on GPU. Are there other problems?

No severe ones. The string_view must be hidden, and then you are using bit2mask from mathutils, but that can be ported to GPU easily.

shahor02 commented on Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx:
```diff
@@ -196,11 +233,15 @@ void TRDGlobalTracking::endOfStream(EndOfStreamContext& ec)
 
-DataProcessorSpec getTRDGlobalTrackingSpec(bool useMC, bool useTrkltTransf)
+DataProcessorSpec getTRDGlobalTrackingSpec(bool useMC, bool useTrkltTransf, o2::dataformats::GlobalTrackID::mask_t src)
 {
```
I think the name assigned to the the device should depend on the input: in the async workflow we will have both instances: using TPC tracks and TPC-ITS and/or global. Could you simply concatenate to "trd-globaltracking" the inputs, e.g.
`o2::utils::concat_string("trd-globaltracking",o2::dataformats::GlobalTrackID::getSourcesNames(src))`

For the same reason, we better define different output names, e.g.
````
outputs.emplace_back(o2::header::gDataOriginTRD,  o2::dataformats::GlobalTrackID::includesDet(DetID::ITS) ?  "MATCHTRD_GLO" : "MATCHTRD_TPC", 0, Lifetime::Timeframe);
````

martenole replied:
Hi @shahor02 is there an advantage of having one instance of the TRD tracker with both inputs (ITS-TPC and TPC) simultaneously? Because technically it is currently possible and if I would allow that than the output name would need to be defined differently, as there could be both `MATHTRD_GLO` and `MATCHTRD_TPC` from the same device. Or if not should I add an error in case both inputs are defined to avoid that mixture?

shahor02 replied:
I meant something else. The idea is to run 1st TPC-TRD matching, which will produce e.g. `MATCHTRD_TPC`.
After similar step in TOF, e.g. inputs TPC and TPC-TRD, the matching to ITS is done for unambigous matches in inputs TPC-TRD-TOF, TPC-TRD, TPC-TOF and TPC, producing either the 4 outputs with ITS added or single output with all ITS-matched tracks mixed (still have to decided which is better, but I have slight preference for single output). Then these outputs should go again to TRD matcher which would match ITS-TPC part of tracks, and,
1) if needed, disable existing TPC-TRD matches not used by ITS-TPC but whose TPC tracks were matched to ITS (in case a greedy matching was done initially for TPC-TRD). By disabling I mean adding approved TPC-TRD tracks to the same output as newly matched ITS-TPC-TRD, so that later the "bad ones" can be filtered out by the RecoContainer logic. 
 2) if initial matching was allowing only unambiguous matches, which were all used in ITS matching as seeds, then this incremental TRD matching should match the rest to improve the eff.

If this is confusing, we can chat on Wed.

shahor02 commented:
@davidrohr actually, I've realized that used `std::bitset` for the GlobalTrackID::mask_t, shall I change it to simple uint32_t?

davidrohr commented:
> @davidrohr actually, I've realized that used `std::bitset` for the GlobalTrackID::mask_t, shall I change it to simple uint32_t?

Actually, there is currently some work ongoing in clang to support bitsets on GPU. Might be that we have it in few days. Would just need that I have to update the GPU container once again...

davidrohr commented:
@martenole : https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx#L109 : Shouldn't this be a fatal error? If I understand correctly, it could lead to a segfault afterwards.

martenole commented:
> @martenole : https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx#L109 : Shouldn't this be a fatal error? If I understand correctly, it could lead to a segfault afterwards.

Absolutely, the program is supposed to abort in that case. Will change it to a fatal error. I am also checking the GPU compilation again with the slc8-gpu-builder container to fix the GPUTRDTrack constructors for the GPU data types that rely on TrackTPC and TrackTPCITS. For TrackTPC I indeed forgot to include the header. But in principle both are not needed and I will hide them for the GPU compilation.
Will also change the device name depending on the input and the same for the output names.

davidrohr commented:
Hi Ole, i would allow both. In case you are processing both, you should create 2 individual outputs, and they can then carry the two names of the single outputs.

Kind Regards
David Rohr

Sent from my mobile. (Excuse the typos!)

On April 26, 2021 8:52:20 AM GMT+02:00, Ole Schmidt ***@***.***> wrote:
***@***.*** commented on this pull request.
>
>
>
>> @@ -196,11 +233,15 @@ void
>TRDGlobalTracking::endOfStream(EndOfStreamContext& ec)
>        mTimer.CpuTime(), mTimer.RealTime(), mTimer.Counter() - 1);
> }
> 
>-DataProcessorSpec getTRDGlobalTrackingSpec(bool useMC, bool
>useTrkltTransf)
>+DataProcessorSpec getTRDGlobalTrackingSpec(bool useMC, bool
>useTrkltTransf, o2::dataformats::GlobalTrackID::mask_t src)
> {
>
>Hi @shahor02 is there an advantage of having one instance of the TRD
>tracker with both inputs (ITS-TPC and TPC) simultaneously? Because
>technically it is currently possible and if I would allow that than the
>output name would need to be defined differently, as there could be
>both `MATHTRD_GLO` and `MATCHTRD_TPC` from the same device. Or if not
>should I add an error in case both inputs are defined to avoid that
>mixture?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/AliceO2Group/AliceO2/pull/5969#discussion_r620016513

davidrohr commented:
@martenole : I have tested this now, and I see 2 issues:
- It always instantiates both the TPC and the ITS-TPC reader, but it should only use those readers that are actually requested. In particular, when I have only TPC tracks, it will fail if the tpcits match root file is not present, since it will create the reader even though not needed.
- Starting the workflow with no options, or with ` --tracking-sources TPC,ITS-TPC` leads to the following crash:
```
[92809:TRDTrackletReader]: [22:23:52][INFO] Pushing 10 TRD trigger records at entry 0
[92809:TRDTrackletReader]: [22:23:52][INFO] Pushing 1029 uncalibrated TRD tracklets for these trigger records
[92809:TRDTrackletReader]: [22:23:52][ERROR] requested channel has not been configured? check channel names/configuration.
[92809:TRDTrackletReader]: [22:23:52][ERROR] channel: from_TRDTrackletReader_to_trd-globaltracking[TPC,ITS-TPC], index: 0
[92809:TRDTrackletReader]: [22:23:52][ERROR] out of range: _Map_base::at
[92809:TRDTrackletReader]: [22:23:52][ERROR] Exception caught: _Map_base::at 
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/O2/dev-local1/lib/libO2FrameworkFoundation.so(_ZN2o29framework13runtime_errorEPKc+0x86)[0x7f471f8e7426]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/O2/dev-local1/lib/libO2Framework.so(+0x117290)[0x7f471fb94290]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/O2/dev-local1/lib/libO2Framework.so(_ZN2o29framework20DataProcessingDevice5doRunERNS0_20DataProcessorContextE+0xcb)[0x7f471fcb26fb]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/O2/dev-local1/lib/libO2Framework.so(_ZN2o29framework20DataProcessingDevice14ConditionalRunEv+0x458)[0x7f471fcb2ec8]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/FairMQ/v1.4.35.1-local1/lib/libFairMQ.so.1.4(_ZN12FairMQDevice10RunWrapperEv+0x405)[0x7f471dd4aa45]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/FairMQ/v1.4.35.1-local1/lib/libFairMQ.so.1.4(+0x1c5e49)[0x7f471dd4ae49]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/FairMQ/v1.4.35.1-local1/lib/libFairMQStateMachine.so.1.4(_ZN5boost6detail8function26void_function_obj_invoker1ISt8functionIFvN4fair2mq5StateEEEvS6_E6invokeERNS1_15function_bufferES6_+0x2d)[0x7f471d9992bd]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/FairMQ/v1.4.35.1-local1/lib/libFairMQStateMachine.so.1.4(_ZN5boost8signals26detail11signal_implIFvN4fair2mq5StateEENS0_19optional_last_valueIvEEiSt4lessIiENS_8functionIS6_EENSB_IFvRKNS0_10connectionES5_EEENS0_5mutexEEclES5_+0x2e9)[0x7f471d9a4cc9]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/FairMQ/v1.4.35.1-local1/lib/libFairMQStateMachine.so.1.4(_ZN4fair2mq3fsm8Machine_11ProcessWorkEv+0x3ff)[0x7f471d9ac4cf]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/FairMQ/v1.4.35.1-local1/lib/libFairMQStateMachine.so.1.4(_ZN4fair2mq12StateMachine11ProcessWorkEv+0x40)[0x7f471d997d90]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/FairMQ/v1.4.35.1-local1/lib/libFairMQ.so.1.4(_ZN4fair2mq12DeviceRunner3RunEv+0x1080)[0x7f471dd14de0]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/O2/dev-local1/lib/libO2Framework.so(_Z7doChildiPPcRN2o29framework15ServiceRegistryERKNS2_19RunningWorkflowInfoENS2_16RunningDeviceRefENS2_17TerminationPolicyERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEP9uv_loop_s+0x25d)[0x7f471fe0e8ed]
[92809:TRDTrackletReader]: /home/qon/alice/sw/gentoo2_x86-64/O2/dev-local1/lib/libO2Framework.so(_Z15runStateMachineRKSt6vectorIN2o29framework17DataProcessorSpecESaIS2_EERK12WorkflowInfoRKS_INS1_17DataProcessorInfoESaISA_EERKNS1_11CommandInfoERNS1_13DriverControlERNS1_10DriverInfoERS_INS1_17DeviceMetricsInfoESaISM_EERN5boost15program_options13variables_mapENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0xb56)[0x7f471fe0f846]
```

martenole commented:
Yes, I should have written that it is not yet ready. I will have a chat with Ruben tomorrow..
About the readers that is a good point, I will fix that now. The second point I also saw of course, but I don't yet have an idea what could be wrong.
And then there is the problem I mentioned at the very beginning about the missing dictionary. When you try the tracking specifying the input to either TPC or ITS-TPC the tracking will work and if you check the outputfile `trdtracks.root` you will see that error. Do you maybe have an idea what could be wrong? When I used the `TrackTPCITS` as base class for the o2 tracks it worked fine, but with `TrackParCov` I have these issues

davidrohr commented:
ok, thx, no problem. I think it makes sense at some point to remove the Draft / WIP from PRs that are not final.
But then I would add something to the title like (STILL TESTING), otherwise people might merge it when it becomes green.

Will have a look at the 2 issues you mention now.

davidrohr commented:
Hi,
could you rename the `tracks-global` to `tracks_global`? I think the `-` confuses ROOT.
If I use as input TPC, it works for me in the TBrowser. I can visualize all the track properties, also the mNTracklets.
Perhaps can you do a rebuild of O2 and try again?
However, running `tracksTRD->Draw("tracks.mP[0]")` it fails for me with
```
root.exe: /var/tmp/portage/sci-physics/root-6.22.06/work/root-6.22.06/interpreter/llvm/src/tools/clang/include/clang/AST/DeclContextInternals.h:119: void clang::StoredDeclsList::remove(clang::NamedDecl*): Assertion `std::find(Vec.begin(), Vec.end(), D) == Vec.end() && "list still contains decl"' failed.
```
No idea what that is, never seen before.

martenole commented:
Hm, renaming to `tracks_global` did not change anything to me. And do the values for `mNTracklets` make sense? Should be something between 0 and 6, but I see a mean of a few millions..
But I don't see an error when trying `tracksTRD->Draw("tracks.mP[0]")`, that works for me

davidrohr commented:
ah, ok, indeed they are bogus. But showing the mP[0 to 5] looked ok.
Weird that our installations behave differently. Can you try with these linkdefs:
```
#pragma link C++ class o2::gpu::trackInterface < o2::track::TrackParametrizationWithError<float>> + ;
#pragma link C++ class o2::gpu::GPUTRDTrack_t < o2::gpu::trackInterface < o2::track::TrackParametrizationWithError<float>>> + ;
#pragma link C++ class std::vector < o2::gpu::GPUTRDTrack_t < o2::gpu::trackInterface < o2::track::TrackParametrizationWithError<float>>>> + ;
```

davidrohr commented:
OK, I found the problem with the 2 inputs. The processor name for the TRD tracker is computed as `o2::utils::concat_string("trd-globaltracking", "GTrackID::getSourcesNames(src)")` and apparently DPL or FairMQ fail if there is a `,` in it. Could you do a string replace from `,` to `_` or so, then it should work.

martenole commented:
> ah, ok, indeed they are bogus. But showing the mP[0 to 5] looked ok.
> Weird that our installations behave differently. Can you try with these linkdefs:
> 
> ```
> #pragma link C++ class o2::gpu::trackInterface < o2::track::TrackParametrizationWithError<float>> + ;
> #pragma link C++ class o2::gpu::GPUTRDTrack_t < o2::gpu::trackInterface < o2::track::TrackParametrizationWithError<float>>> + ;
> #pragma link C++ class std::vector < o2::gpu::GPUTRDTrack_t < o2::gpu::trackInterface < o2::track::TrackParametrizationWithError<float>>>> + ;
> ```

Still no luck. I managed to avoid the ROOT error `Error in <TInterpreter::AutoParse>: Error parsing payload code for class` by including GPUTRDDef.h in my interactive ROOT session before trying to draw the tracks.
But still I have the impression that something goes very wrong with something (the ROOT interpreter maybe?)

I try an interactive ROOT session:

root [0] o2::gpu::GPUTRDTrack trk
ROOT_prompt_0:1:10: error: no member named 'GPUTRDTrack' in namespace 'o2::gpu'
o2::gpu::GPUTRDTrack trk
~~~~~~~~~^
root [1] #include "GPUTRDDef.h"
root [2] o2::gpu::GPUTRDTrack trk
(o2::gpu::GPUTRDTrack &) @0x7fc58f51d010
root [3] trk.GetNtrackletsOffline()
ROOT_prompt_3:1:26: error: too few arguments to function call, single argument 'type' was not specified
trk.GetNtrackletsOffline()
~~~~~~~~~~~~~~~~~~~~~~~~ ^
G__O2GPUTracking dictionary payload:1197:10: note: 'GetNtrackletsOffline' declared here
  GPUd() int GetNtrackletsOffline(int type) const { return mNTrackletsOffline[type]; }
         ^
root [4] trk.GetNtrackletsOffline(0)
ROOT_prompt_4:1:5: error: no member named 'GetNtrackletsOffline' in 'o2::gpu::GPUTRDTrack_t<o2::gpu::trackInterface<o2::track::TrackParametrizationWithError<float> > >'
trk.GetNtrackletsOffline(0)
~~~ ^
root [5] trk.GetNtracklets()
(int) 0
root [6]

martenole commented:
So what is wrong is still the number of tracklets which is nonsense for the calculated TRD tracks

davidrohr commented:
Hi Ole,

can you create a stupid dummy class
```
namespace o2::gpu {
class O2GPUTRDBase : public o2::track::TrackParCov {
// you might need to forward non-default constructors here, if you use them
};
}
```

and use that as the base class. Then it should behave identically compared to the TrackTPCITS. I still suspect that perhaps there is a problem with the ROOT parsing of the template and the `using TrackParCov = ...`.

martenole commented:
Finally now with the dummy class the dictionary generation seems to work :)
The nonsense values for the number of tracklets on the other hand were caused by some errors in the `GPUTRDTrack` constructors. Although I am not quite sure why. I delegated the constructors from different input types to the default constructor and that apparently did not work. Instead I added now an `Initialize()` method which sets the class members to the default values and call that from every constructor.
Thanks a lot for the help!

davidrohr commented:
Hi Ole, can you then squash the other commits, but keep a separate commit for the dummy class, and keep a comment in the dummy class header, that this can be reverted once this problem in ROOT is fixed.
Actually, I am using root 6.22 myself, and I didn't see this dictionary issue anymore and 6.24 has just come out and we will bump soon. So hopefully we can also revert it soon :)

martenole committed:
Add dummy TRD base class. Probably to be reverted

After ROOT bump to 6.22 or newer. Not clear why, but this is a workaround for
an issue with the ROOT interpreter which is not seen anymore in ROOT 6.22.
More details in PR 5969

martenole commented:
Done, waiting for the CI now and then this is in principle ready to be merged.

martenole commented:
Btw I tried reverting the last commit again, but then I did see the issue with the interpreter again. But only when using the command line. When I was working with the TBrowser everything seemed to be fine already.

davidrohr commented:
OK, let's get this merged. Then we can actually try to create a small test case and file this as bug report to root.
